pkg/s3store/s3store.go

```diff
@@ -272,6 +277,70 @@ func (upload *s3Upload) writeInfo(ctx context.Context, info handler.FileInfo) er
 	return err
 }
 
+type s3ChunkProducer struct {
+	files chan<- *os.File
+	done  chan struct{}
+	err   error
+	r     io.Reader
+}
+
+func (scp *s3ChunkProducer) produce(chunkSize int64) {
+	for {
+		file, err := scp.nextChunk(chunkSize)
+		if err != nil {
+			scp.err = err
+			close(scp.files)
+			return
+		}
+		if file == nil {
+			close(scp.files)
+			return
+		}
+		select {
+		case scp.files <- file:
+		case <-scp.done:
+			close(scp.files)
+			return
+		}
+	}
+}
+
+func (scp *s3ChunkProducer) nextChunk(size int64) (*os.File, error) {
+	// Create a temporary file to store the part
+	file, err := ioutil.TempFile("", "tusd-s3-tmp-")
+	if err != nil {
+		return nil, err
+	}
+
+	limitedReader := io.LimitReader(scp.r, size)
+	n, err := io.Copy(file, limitedReader)
+
+	// If the HTTP PATCH request gets interrupted in the middle (e.g. because
+	// the user wants to pause the upload), Go's net/http returns an io.ErrUnexpectedEOF.
+	// However, for S3Store it's not important whether the stream has ended
+	// on purpose or accidentally. Therefore, we ignore this error to not
+	// prevent the remaining chunk to be stored on S3.
+	if err == io.ErrUnexpectedEOF {
+		err = nil
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	// io.Copy returns (0, nil) when it reaches EOF
```

Comment:
I am aware that this comment has already been there before your PR, but I think it is wrong. io.Copy returns (0, nil) if it is reaches EOF *and* did not read an bytes (i.e. the input buffer is empty).

I think in that case, a better comment would be:
> If the entire request body is read and no more data is available, io.Copy returns 0 since it is unable to read any bytes. In that case, we can close the s3PartProducer.

What do you think?